File: algorithms/recursive/tree.js
// DIY tree utility in Node.js
// Andrew Cashner
// 2026/07/15

import * as fs from "node:fs";
import * as path from "node:path";

function showfile(filename, indent) {
  if (indent == 0) {
    console.log(filename);
  } else {
    let preface = "  ".repeat(indent - 1);
    console.log(`${preface}|__ ${filename}`);
  }

}

function display_dir(filepath, indent = 0) {
  showfile(path.basename(filepath), indent);

  let pathStatus = fs.statSync(filepath, { throwIfNoEntry: false });

  if (pathStatus && pathStatus.isDirectory()) {
      let files = fs.readdirSync(filepath, { withFileTypes: true });
      if (files) {
        files.forEach(file => display_dir(`${filepath}/${file.name}`, indent + 1));
      }
  }
}

function main() {
  if (process.argv.length != 3) {
    console.log("Usage: node tree.js PATH");
  } else {
    let path = process.argv[2];
    display_dir(path);
  }
}

main();

File: algorithms/recursive/tree/tree.cpp
/* DIY tree utility (C++)
 *
 * Andrew Cashner
 * 2026/07/16
 */

#include <iostream>
#include <string>
#include <filesystem>

void print_dir(std::string path);

int main(int argc, char *argv[]) 
{
    if (argc != 2)
    {
        std::cerr << "Usage: tree PATH\n";
        std::exit(EXIT_FAILURE);
    }

    std::string path { argv[1] };
    print_dir(path);

    return 0;
}

void print_dir(std::string path)
{
    std::cout << path << "\n";

    auto iterator = std::filesystem::recursive_directory_iterator(path);

    for (const std::filesystem::directory_entry& entry : iterator)
    {
        std::string indent = std::string(2 * iterator.depth(), ' ');
        std::string basename { entry.path().filename() };

        std::cout << indent << "|__ " << basename << "\n";
    }
}

